File: server/src/libs/ocr.py
"""
ocr.py - OCR 文字识别（getWords / findStr）

模式：
  默认模式: stdin 接收 JSON，stdout 输出 JSON（一次性）
  --worker: 持久化工作进程，逐行读入 JSON 命令，逐行输出 JSON 结果

依赖: opencv-python, numpy, rapidocr-onnxruntime
"""

import sys
import json
import time
import base64
import numpy as np
import cv2
from rapidocr_onnxruntime import RapidOCR

# 全局 OCR 引擎（复用，避免重复加载模型）
_ocr_engine = None


def get_ocr():
    global _ocr_engine
    if _ocr_engine is None:
        _ocr_engine = RapidOCR(params={"Global.use_cls": False})
    return _ocr_engine


# ── 预设颜色（HSV 范围，OpenCV 色相 0-180） ────

COLOR_PRESETS = {
    "red": [(0, 50, 50), (10, 255, 255), (160, 50, 50), (180, 255, 255)],
    "orange": [(10, 50, 50), (25, 255, 255)],
    "yellow": [(25, 50, 50), (35, 255, 255)],
    "green": [(35, 50, 50), (85, 255, 255)],
    "cyan": [(85, 50, 50), (100, 255, 255)],
    "blue": [(100, 50, 50), (130, 255, 255)],
    "purple": [(130, 50, 50), (160, 255, 255)],
    "white": [(0, 0, 200), (180, 40, 255)],
    "black": [(0, 0, 0), (180, 255, 50)],
    "gray": [(0, 0, 50), (180, 40, 200)],
}


def decode_base64_image(b64_string: str) -> np.ndarray:
    """将 base64 字符串解码为 OpenCV 图像（BGR）"""
    if "," in b64_string:
        b64_string = b64_string.split(",", 1)[1]
    raw = base64.b64decode(b64_string)
    arr = np.frombuffer(raw, dtype=np.uint8)
    img = cv2.imdecode(arr, cv2.IMREAD_COLOR)
    if img is None:
        raise ValueError("无法解码图片")
    return img


def parse_color_to_bgr(color: str):
    """将颜色字符串解析为 BGR 元组。
    支持：
      - 预设名称 (red, blue, ...)
      - hex (#RRGGBB 或 #RGB)
      - rgb (r,g,b)
    """
    s = color.strip()
    # hex
    if s.startswith('#'):
        h = s.lstrip('#')
        if len(h) == 3:
            h = ''.join(c * 2 for c in h)
        if len(h) != 6:
            raise ValueError(f"无效的 hex 颜色: {color}")
        r, g, b = int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16)
        return (b, g, r)
    # rgb comma-separated
    parts = [p.strip() for p in s.split(',')]
    if len(parts) == 3:
        r, g, b = int(parts[0]), int(parts[1]), int(parts[2])
        return (b, g, r)
    raise ValueError(f"无法解析颜色: {color}")


def apply_color_filter(img: np.ndarray, color: str, color_tolerance: int = 50) -> np.ndarray:
    """按颜色过滤：保留指定颜色区域的像素，其余设为白色。
    color_tolerance: 0-100，默认 50。映射为 HSV 容差：色相 = tol*0.3，饱和度/明度 = tol*1.2。
    """
    color_lower = color.lower().strip()
    h_tol = max(1, int(color_tolerance * 0.3))
    sv_tol = max(1, int(color_tolerance * 1.2))

    # preset name
    if color_lower in COLOR_PRESETS:
        hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
        ranges = COLOR_PRESETS[color_lower]
        # scale preset ranges by tolerance factor (default 50 => factor 1.0)
        factor = color_tolerance / 50.0
        if len(ranges) == 4:
            l1, u1, l2, u2 = ranges
            def scale_range(lo, hi):
                mid = [(lo[i] + hi[i]) / 2 for i in range(3)]
                half = [(hi[i] - lo[i]) / 2 * factor for i in range(3)]
                s_lo = [max(0 if i == 0 else 0, int(mid[i] - half[i])) for i in range(3)]
                s_hi = [min(180 if i == 0 else 255, int(mid[i] + half[i])) for i in range(3)]
                return tuple(s_lo), tuple(s_hi)
            sl1, sh1 = scale_range(l1, u1)
            sl2, sh2 = scale_range(l2, u2)
            mask1 = cv2.inRange(hsv, np.array(sl1), np.array(sh1))
            mask2 = cv2.inRange(hsv, np.array(sl2), np.array(sh2))
            mask = cv2.bitwise_or(mask1, mask2)
        else:
            lo, hi = ranges[0], ranges[1]
            mid = [(lo[i] + hi[i]) / 2 for i in range(3)]
            half = [(hi[i] - lo[i]) / 2 * factor for i in range(3)]
            s_lo = tuple(max(0, int(mid[i] - half[i])) for i in range(3))
            s_hi = tuple(min(180 if i == 0 else 255, int(mid[i] + half[i])) for i in range(3))
            mask = cv2.inRange(hsv, np.array(s_lo), np.array(s_hi))
    else:
        # arbitrary hex / rgb -> convert to HSV, build tolerance range
        bgr = parse_color_to_bgr(color)
        pixel = np.uint8([[list(bgr)]])
        hsv_pixel = cv2.cvtColor(pixel, cv2.COLOR_BGR2HSV)
        h, s, v = int(hsv_pixel[0][0][0]), int(hsv_pixel[0][0][1]), int(hsv_pixel[0][0][2])
        # tolerance derived from color_tolerance parameter
        lower = np.array([max(0, h - h_tol), max(0, s - sv_tol), max(0, v - sv_tol)])
        upper = np.array([min(180, h + h_tol), min(255, s + sv_tol), min(255, v + sv_tol)])
        hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
        mask = cv2.inRange(hsv, lower, upper)

    result = np.full_like(img, 255)
    result[mask > 0] = img[mask > 0]
    gray = cv2.cvtColor(result, cv2.COLOR_BGR2GRAY)
    return cv2.cvtColor(gray, cv2.COLOR_GRAY2BGR)


def get_words(image_b64: str, region: list = None, color: str = None,
             color_tolerance: int = 50) -> dict:
    """OCR 文字识别"""
    t0 = time.perf_counter()
    src = decode_base64_image(image_b64)

    offset_x, offset_y = 0, 0
    if region and len(region) == 4:
        x1, y1, x2, y2 = region
        if x1 != 0 or y1 != 0 or x2 != 0 or y2 != 0:
            x1 = max(0, min(x1, src.shape[1]))
            y1 = max(0, min(y1, src.shape[0]))
            x2 = max(x1 + 1, min(x2, src.shape[1]))
            y2 = max(y1 + 1, min(y2, src.shape[0]))
            src = src[y1:y2, x1:x2]
            offset_x, offset_y = x1, y1

    if color:
        src = apply_color_filter(src, color, color_tolerance)

    ocr = get_ocr()
    result, elapse = ocr(src)

    words = []
    if result:
        for item in result:
            box, text, confidence = item
            xs = [p[0] for p in box]
            ys = [p[1] for p in box]
            bx = int(min(xs))
            by = int(min(ys))
            bw = int(max(xs) - bx)
            bh = int(max(ys) - by)
            words.append({
                "text": text,
                "x": bx + offset_x,
                "y": by + offset_y,
                "w": bw,
                "h": bh,
                "confidence": round(float(confidence), 4),
            })

    elapsed = round((time.perf_counter() - t0) * 1000, 1)
    return {"words": words, "elapsed": elapsed}


def levenshtein_ratio(s1: str, s2: str) -> float:
    """计算两个字符串的相似度（基于编辑距离），返回 0-1"""
    if not s1 and not s2:
        return 1.0
    m, n = len(s1), len(s2)
    dp = list(range(n + 1))
    for i in range(1, m + 1):
        prev = dp[0]
        dp[0] = i
        for j in range(1, n + 1):
            temp = dp[j]
            if s1[i - 1] == s2[j - 1]:
                dp[j] = prev
            else:
                dp[j] = 1 + min(prev, dp[j], dp[j - 1])
            prev = temp
    dist = dp[n]
    max_len = max(m, n)
    return 1.0 - dist / max_len if max_len > 0 else 0.0


def sliding_window_match(text: str, target: str) -> float:
    """滑动窗口匹配：在 text 中查找与 target 最相似的子串"""
    t_len = len(target)
    if t_len == 0:
        return 0.0
    if target in text:
        return 1.0
    best = 0.0
    for window_size in range(max(1, t_len - 1), min(len(text), t_len + 1) + 1):
        for i in range(len(text) - window_size + 1):
            substring = text[i:i + window_size]
            ratio = levenshtein_ratio(substring, target)
            if ratio > best:
                best = ratio
            if best >= 1.0:
                return best
    return best


def find_str(image_b64: str, target: str, region: list = None,
             similarity: float = 0.8, color: str = None,
             color_tolerance: int = 50) -> dict:
    """找字：在截图中查找指定文字"""
    ocr_result = get_words(image_b64, region, color, color_tolerance)
    all_words = ocr_result["words"]

    matches = []
    for word in all_words:
        text = word["text"]
        if target in text:
            matches.append({**word, "similarity": 1.0})
            continue
        sim = sliding_window_match(text, target)
        if sim >= similarity:
            matches.append({**word, "similarity": round(sim, 4)})

    matches.sort(key=lambda m: m["similarity"], reverse=True)
    return {
        "matches": matches,
        "allWords": [w["text"] for w in all_words],
        "elapsed": ocr_result["elapsed"],
    }


def handle_command(input_data: dict) -> dict:
    """处理单个命令"""
    action = input_data.get("action", "")
    image_b64 = input_data.get("image", "")
    region = input_data.get("region", [0, 0, 0, 0])
    color = input_data.get("color")
    color_tolerance = input_data.get("colorTolerance", 50)

    if not image_b64:
        return {"error": "image 为必填"}

    try:
        if action == "getWords":
            return get_words(image_b64, region, color, color_tolerance)
        elif action == "findStr":
            target = input_data.get("target", "")
            similarity = input_data.get("similarity", 0.8)
            if not target:
                return {"error": "target 为必填"}
            return find_str(image_b64, target, region, similarity, color, color_tolerance)
        else:
            return {"error": f"未知 action: {action}，支持: getWords, findStr"}
    except Exception as e:
        return {"error": str(e)}


def worker_loop():
    """持久化工作进程模式：预加载模型，逐行读入命令"""
    # 预热：加载 OCR 模型（首次调用最慢，后续复用）
    get_ocr()
    # 输出就绪信号
    sys.stdout.write(json.dumps({"ready": True}) + "\n")
    sys.stdout.flush()

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            input_data = json.loads(line)
            if input_data.get("action") == "exit":
                break
            result = handle_command(input_data)
        except json.JSONDecodeError as e:
            result = {"error": f"JSON 解析失败: {e}"}
        except Exception as e:
            result = {"error": str(e)}

        sys.stdout.write(json.dumps(result, ensure_ascii=False) + "\n")
        sys.stdout.flush()


def main():
    """一次性模式：stdin 读取全部 JSON，处理后输出"""
    try:
        input_data = json.loads(sys.stdin.read())
    except json.JSONDecodeError as e:
        print(json.dumps({"error": f"JSON 解析失败: {e}"}))
        sys.exit(1)

    result = handle_command(input_data)
    print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    if "--worker" in sys.argv:
        worker_loop()
    else:
        main()

File: server/src/libs/ocr.ts
// ─────────────────────────────────────────────
// ocr — OCR 文字识别核心 API
//
// 提供 getWords() 和 findStr() 两个接口，
// 底层通过 Python + RapidOCR 实现。
//
// 使用方：import { getWords, findStr } from '../libs/index.js'
// ─────────────────────────────────────────────

import { PythonWorker } from './python-helper.js'
import { logger } from '../core/logger.js'

// ── 类型定义 ──────────────────────────────────

/** 预设颜色名称 */
export type OcrColorPreset =
  'red' | 'orange' | 'yellow' | 'green' | 'cyan' | 'blue' | 'purple' | 'white' | 'black' | 'gray'

/** 颜色：预设名称 或 任意 hex (#RRGGBB / #RGB) 或 rgb (r,g,b) */
export type OcrColor = OcrColorPreset | (string & {})

/** OCR 识别到的单个文字块 */
export interface OcrWord {
  /** 识别出的文字 */
  text: string
  /** 左上角 X（原始截图分辨率） */
  x: number
  /** 左上角 Y（原始截图分辨率） */
  y: number
  /** 文字块宽度 */
  w: number
  /** 文字块高度 */
  h: number
  /** 置信度 0-1 */
  confidence: number
}

/** getWords 输入参数 */
export interface GetWordsOptions {
  /** 截图 base64（支持 data:image/png;base64,... 或纯 base64） */
  image: string
  /** 识别区域 [x1, y1, x2, y2]，全为 0 或不传表示全图 */
  region?: [number, number, number, number]
  /** 文字颜色过滤，不传则识别所有颜色 */
  color?: OcrColor
  /** 颜色偏差 0-100，默认 50。值越大容差越宽 */
  colorTolerance?: number
}

/** getWords 结果 */
export interface GetWordsResult {
  /** 识别到的文字列表 */
  words: OcrWord[]
  /** 总耗时（毫秒） */
  elapsed: number
}

/** findStr 匹配结果（在 OcrWord 基础上增加相似度） */
export interface OcrMatch extends OcrWord {
  /** 文字匹配相似度 0-1 */
  similarity: number
}

/** findStr 输入参数 */
export interface FindStrOptions {
  /** 截图 base64 */
  image: string
  /** 要查找的文字 */
  target: string
  /** 查找区域 [x1, y1, x2, y2]，全为 0 或不传表示全图 */
  region?: [number, number, number, number]
  /** 相似度阈值 0-1，默认 0.8 */
  similarity?: number
  /** 文字颜色过滤 */
  color?: OcrColor
  /** 颜色偏差 0-100，默认 50。值越大容差越宽 */
  colorTolerance?: number
}

/** findStr 结果 */
export interface FindStrResult {
  /** 匹配结果列表（按相似度降序） */
  matches: OcrMatch[]
  /** OCR 识别到的所有文字（调试用） */
  allWords: string[]
  /** 总耗时（毫秒） */
  elapsed: number
}

/** Python 输出格式（内部） */
interface OcrPyOutput {
  words?: OcrWord[]
  matches?: OcrMatch[]
  allWords?: string[]
  elapsed?: number
  error?: string
  [key: string]: unknown
}

// ── 公开函数 ──────────────────────────────────

/** OCR 持久化工作进程（模型只加载一次，后续调用直接通信） */
const ocrWorker = new PythonWorker('ocr.py')

/** 销毁 OCR 工作进程（服务关闭时调用，防止 python 进程残留） */
export function destroyOcrWorker(): void {
  ocrWorker.destroy()
}

/**
 * 获取文字（OCR 识别）
 *
 * 识别截图中的所有文字，支持区域裁切和颜色过滤。
 *
 * @example
 * ```ts
 * import { getWords } from '../libs/index.js'
 *
 * const result = await getWords({
 *   image: screenshotBase64,
 *   region: [100, 50, 800, 600],
 *   color: 'red',
 * })
 *
 * for (const word of result.words) {
 *   console.log(`[${word.x},${word.y}] ${word.text} (${word.confidence})`)
 * }
 * ```
 */
export async function getWords(options: GetWordsOptions): Promise<GetWordsResult> {
  const { image, region = [0, 0, 0, 0], color, colorTolerance = 50 } = options

  if (!image) {
    throw new Error('image 为必填')
  }

  const t0 = Date.now()

  const pyResult = await ocrWorker.send<OcrPyOutput>({
    action: 'getWords',
    image,
    region,
    color: color ?? null,
    colorTolerance,
  })

  if (pyResult.error) {
    logger.error('getWords', `底层错误: ${pyResult.error}`)
    throw new Error(pyResult.error)
  }

  const words: OcrWord[] = (pyResult.words ?? []).map((w) => ({
    text: w.text,
    x: w.x,
    y: w.y,
    w: w.w,
    h: w.h,
    confidence: w.confidence,
  }))

  const elapsed = Date.now() - t0
  logger.info('getWords', `${words.length} words, py=${pyResult.elapsed}ms, total=${elapsed}ms`)

  return { words, elapsed }
}

/**
 * 找字（OCR + 文字匹配）
 *
 * 在截图中查找指定文字，支持区域裁切、颜色过滤和相似度控制。
 *
 * @example
 * ```ts
 * import { findStr } from '../libs/index.js'
 *
 * const result = await findStr({
 *   image: screenshotBase64,
 *   target: '确认',
 *   similarity: 0.9,
 * })
 *
 * if (result.matches.length > 0) {
 *   const { x, y, w, h } = result.matches[0]
 *   console.log(`找到文字 "${result.matches[0].text}" 在 (${x}, ${y})`)
 * }
 * ```
 */
export async function findStr(options: FindStrOptions): Promise<FindStrResult> {
  const {
    image,
    target,
    region = [0, 0, 0, 0],
    similarity = 0.8,
    color,
    colorTolerance = 50,
  } = options

  if (!image || !target) {
    throw new Error('image 和 target 均为必填')
  }

  const t0 = Date.now()

  const pyResult = await ocrWorker.send<OcrPyOutput>({
    action: 'findStr',
    image,
    target,
    region,
    similarity,
    color: color ?? null,
    colorTolerance,
  })

  if (pyResult.error) {
    logger.error('findStr', `底层错误: ${pyResult.error}`)
    throw new Error(pyResult.error)
  }

  const matches: OcrMatch[] = (pyResult.matches ?? []).map((m) => ({
    text: m.text,
    x: m.x,
    y: m.y,
    w: m.w,
    h: m.h,
    confidence: m.confidence,
    similarity: m.similarity,
  }))

  const elapsed = Date.now() - t0
  logger.info('findStr', `${matches.length} matches, py=${pyResult.elapsed}ms, total=${elapsed}ms`)
  if (matches.length === 0) {
    logger.warn('findStr', `OCR 识别到的文字: ${JSON.stringify(pyResult.allWords ?? [])}`)
  }

  return { matches, allWords: pyResult.allWords ?? [], elapsed }
}
